File: python/day6.py
DIRS = [
    (0, -1),  # ^
    (1, 0),  # >
    (0, 1),  # V
    (-1, 0),  # <
]


def solve_part1(map):
    x, y = find_guard(map)
    width, height = len(map[0]), len(map)
    visited = set()

    dir = 0
    count = 0
    while 0 <= x < width and 0 <= y < height:
        if map[y][x] == "#":
            dx, dy = DIRS[dir]
            x, y = x - dx, y - dy
            dir = (dir + 1) % 4
        elif (x,y) not in visited:
            count += 1

        visited.add((x,y))
        dx, dy = DIRS[dir]
        x, y = x + dx, y + dy

    return count, visited


def solve_part2(map):
    # Map out the guard's path first so that we can focus on putting obstacles
    # where they'll affect the pathfinding, avoiding unnecessary work
    _, walked = solve_part1(map)

    gx, gy = find_guard(map)
    map[gy][gx] = "."  # easier to blank the guard marker to avoid more checks

    width, height = len(map[0]), len(map)
    count = 0
    for y in range(height):
        for x in range(width):
            if gy == y and gx == x or (x,y) not in walked:
                continue
            prev = map[y][x]
            map[y][x] = "O"
            looped = walk(map, gx, gy)
            count += 1 if looped else 0
            map[y][x] = prev

    return count


def find_guard(map):
    for y in range(len(map)):
        for x in range(len(map[0])):
            if map[y][x] == "^":
                return x, y
    return 0, 0


# A modified version of part 1 which exits early as soon as a loop is detected
def walk(map, x, y):
    width, height = len(map[0]), len(map)
    dir = 0
    visited = {}
    while 0 <= x < width and 0 <= y < height:
        current = visited.get((x, y), map[y][x])
        dx, dy = DIRS[dir]

        if current in "#O":
            x, y = x - dx, y - dy
            dir = (dir + 1) % 4
            dx, dy = DIRS[dir]
        elif current == "+" and map[y + dy][x + dx] in ["#", "O"]:
            # Hit a barrier and we've walked here twice already => looped
            return True

        visited[(x, y)] = ("|" if dir in (0, 2) else "-") if current == "." else "+"
        y, x = y + dy, x + dx

    return False


if __name__ == "__main__":
    with open("day6.input", "r") as f:
        input = f.read().strip("\n")

    map = [list(row) for row in input.splitlines()]
    print("Part 1:", solve_part1(map)[0])
    print("Part 2:", solve_part2(map))

File: zig/day6.zig
const std = @import("std");

const Vec = struct {
    const Self = @This();

    x: isize,
    y: isize,

    inline fn init(x: anytype, y: anytype) Self {
        return .{ .x = @intCast(x), .y = @intCast(y) };
    }

    inline fn toIndex(self: Self, width: usize) usize {
        const iwidth: isize = @intCast(width);
        return @intCast(self.y * iwidth + self.x + self.y); // add y for newlines
    }

    inline fn fromIndex(index: usize, width: usize) Self {
        return Self.init(index % (width + 1), index / (width + 1)); // add 1 for newlines
    }

    inline fn add(self: Self, other: Self) Self {
        return .{ .x = self.x + other.x, .y = self.y + other.y };
    }

    inline fn sub(self: Self, other: Self) Self {
        return .{ .x = self.x - other.x, .y = self.y - other.y };
    }

    inline fn eql(self: Self, other: Self) bool {
        return self.x == other.x and self.y == other.y;
    }
};

const DIRS = [_]Vec{
    .{ .x = 0, .y = -1 }, // ^
    .{ .x = 1, .y = 0 }, // >
    .{ .x = 0, .y = 1 }, // V
    .{ .x = -1, .y = 0 }, // <
};
const BACKGROUND = '.';
const BARRIER = '#';
const OBSTACLE = 'O';

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    const input = try std.fs.cwd().readFileAlloc(allocator, "day6.input", std.math.maxInt(usize));

    const trimmed = std.mem.trim(u8, input, "\n");
    const width = std.mem.indexOfScalar(u8, input, '\n').?;
    const height = std.mem.count(u8, trimmed, "\n") + 1;

    const p1, _ = try solvePart1(allocator, input, width, height);
    std.debug.print("Part 1: {d}\n", .{p1});

    var timer = try std.time.Timer.start();
    timer.reset();
    const result = try solvePart2(allocator, input, width, height);
    const elapsed_ns = timer.read();
    const elapsed = @as(f64, @floatFromInt(elapsed_ns)) / std.time.ns_per_s;
    std.debug.print("Part 2: {d:<3}  =>  {d:.6} seconds\n", .{ result, elapsed });
}

fn solvePart1(allocator: std.mem.Allocator, map: []const u8, width: usize, height: usize) !struct { usize, std.AutoHashMap(Vec, void) } {
    var guard = findGuard(map, width);
    var visited = std.AutoHashMap(Vec, void).init(allocator);
    var sum: usize = 0;
    var dir: usize = 0;
    while (0 <= guard.x and guard.x < width and 0 <= guard.y and guard.y < height) {
        if (map[guard.toIndex(width)] == BARRIER) {
            guard = guard.sub(DIRS[dir]);
            dir = (dir + 1) % 4;
        } else if (!visited.contains(guard)) {
            sum += 1;
        }

        try visited.put(guard, {});
        guard = guard.add(DIRS[dir]);
    }
    return .{ sum, visited };
}

fn solvePart2(allocator: std.mem.Allocator, map: []u8, width: usize, height: usize) !usize {
    _, const walked = try solvePart1(allocator, map, width, height);

    const guard = findGuard(map, width);
    map[guard.toIndex(width)] = BACKGROUND;

    var visited = std.AutoHashMap(Vec, u8).init(allocator);
    try visited.ensureTotalCapacity(8192);
    var sum: usize = 0;
    for (0..height) |y| {
        for (0..width) |x| {
            const pos = Vec.init(x, y);
            if (guard.eql(pos) or !walked.contains(pos)) continue;
            const i = pos.toIndex(width);
            const prev = map[i];
            map[i] = OBSTACLE;
            sum += @intFromBool(try walk(map, width, height, guard, &visited));
            visited.clearRetainingCapacity();
            map[i] = prev;
        }
    }
    return sum;
}

fn findGuard(map: []const u8, width: usize) Vec {
    const offset = std.mem.indexOfScalar(u8, map, '^').?;
    return Vec.fromIndex(offset, width);
}

// A modified version of part 1 which exits early as soon as a loop is detected
fn walk(map: []const u8, width: usize, height: usize, guard: Vec, visited: *std.AutoHashMap(Vec, u8)) !bool {
    var g = guard;
    var dir: usize = 0;
    while (0 <= g.x and g.x < width and 0 <= g.y and g.y < height) {
        var d = DIRS[dir];
        const current = try visited.getOrPutValue(g, map[g.toIndex(width)]);
        const value: u8 = current.value_ptr.*;

        if (value == BARRIER or value == OBSTACLE) {
            g = g.sub(d);
            dir = (dir + 1) % 4;
            d = DIRS[dir];
        } else if (value == '+') {
            const next = map[g.add(d).toIndex(width)];
            if (next == BARRIER or next == OBSTACLE) return true;
        }

        try visited.put(g, if (value == BACKGROUND) (if (dir == 0 or dir == 2) '|' else '-') else '+');
        g = g.add(d);
    }

    return false;
}
